### DISCUSSION TITLE
Can a Failed Job Lead to Database Issues?

### DISCUSSION THREAD

**Discussion by bobwurtz** [stefanzweifel/sidecar-browsershot#125] [Q&A]
Hello,

I recently experienced an issue and I don't know if SidecarBrowsershot is the cause, but I figure it's at least worth asking. Yesterday morning, I started receiving strange timeout errors on my site. It took me until this morning to find the cause. The issue was with availability of connections to my database (screenshot below).

![Screenshot 2024-04-09 at 1 50 42 PM](https://github.com/stefanzweifel/sidecar-browsershot/assets/59666225/e29b31dd-64a9-474c-9702-6b470c9ecf9b)

Right around when these errors started, I started receiving the following error message related to browsershot:

```
Lambda Execution Exception for Wnx\SidecarBrowsershot\Functions\BrowsershotFunction: "Command failed: node ./browser.cjs '-f file:///tmp/browsershot.js'
{ type: 'ElementNotFound' }
. [TRACE] Error: Command failed: node ./browser.cjs '-f file:///tmp/browsershot.js' { type: 'ElementNotFound' } ".
```

My questions are:
- Are these two issues related?
- Could failed browsershot jobs have caused a limitation on the connections to my database?
- Or is it the other way around - did the lack of database connections cause these browsershot jobs to fail?
- Or - are these two things completely unrelated?

I realize this is a strange question, so any feedback/comments/questions are welcome. Thank you

**Comment by stefanzweifel** [comment]
Hi @bobwurtz 

I don't think this is directly related to sidecar-browsershot. The package doen't contain code to interact with a database so it never establishes a connection that could lead to a "too many connections" error.

However, as you mention that you run sidecar-browsershot in a Job, this could potentially dead lock your database:

Let's imagine you're dispatching 50 jobs. Each job triggers sidecar-browsershot. For unkown reasons all sidecar-browsershot executions fail after 10 seconds (maybe the timeout you've set up).

If you're queue setup runs the 50 jobs in parallel (meaning you have 50 workers), all 50 jobs will fail at the same time. Laravel will then attempt to write 50 new rows into your `failed_jobs`-table. Depending on your MySQL settings, this could lead to the error you've shared.

---

I don't know your exact setup, so I can't share a solution directly, but here are some further questions that could help resolve the issue:

- How many jobs running sidecar-browsershot can run in parallel?
- Should there be a rate limit, on how many jobs can run in a given timeframe? (See code below)
- Should job execution maybe be delayed by a random number of seconds, if this error can occur multiple times?

---

I've used this pattern in the past to limit the number of mails a project can send per minute, by using a custom rate limit and using the `RateLimited`-middleware in my job.

This is the code for my custom rate limit. In my case `$job->yourModel` represented a "Newsletter Issue" for a given day. Meaning that only 60 mails per minute were sent for a given newsletter issue. 

```php
RateLimiter::for('sidecar-browsershot-limit', function (ShouldQueue $job) {
    return Limit::perMinute(60)->by($job->yourModel->id);
});
```

In my job I added a middleware method and referenced the rate limit.

```php
public function middleware()
{
    return [
        new \Illuminate\Queue\Middleware\RateLimited('sidecar-browsershot-limit'),
    ];
}
```

**Reply by bobwurtz** [reply]
@stefanzweifel I appreciate the response. I didn't think they were related but had to ask. I already attempt to implement rate limiting but I will look at it again. Thank you!
